cdap-data-fabric/src/main/java/io/cdap/cdap/data2/dataset2/lib/cube/DefaultCube.java

```diff
@@ -458,20 +461,26 @@ private void incrementMetric(String metricName, long value) {
         Collections.sort(timeValues);
         List<TimeValue> resultTimeValues = new ArrayList<>();
 
-        MetricsAggregationOption aggregationOption = query.getAggregationOption();
-        // only do aggregation if the data points are larger than the required count
-        if (aggregationOption != null && aggregationOption.getCount() < timeValues.size()) {
-          int partitionSize = timeValues.size() / aggregationOption.getCount();
+        AggregationOption aggregationOption = query.getAggregationOption();
+        // this should not happen in production, since the check has been made in the handler
+        if (query.getLimit() <= 0) {
+          throw new IllegalArgumentException("The query limit cannot be less than 0");
+        }
+        // only partition the data points if the data points are larger than the required limit and only do it for
+        // option LATEST and SUM.
+        if (query.getLimit() < timeValues.size() && PARTITION_AGG_OPTIONS.contains(aggregationOption)) {
+          int partitionSize = timeValues.size() / query.getLimit();
```

Comment:
Please see my suggestion in https://issues.cask.co/browse/CDAP-15515. 
If you change this to (timeValues.size() + query.getLimit() - 1)/query.getLimit(), then it is rounds up instead of down.

Reply:
I think doing this round up or down will not satisfy all use cases. 
For example, if the number of data points is 100, and count is 70, then the partition size is 2(round up), which will only give 50 data points back. 
For round down, if number of data points is 100, and count is 70, we will return 100 data points in current implementation. 
I will add a commit to return the exact number of data points. Its logic is as follows: Depending on whether there is a remainder R, the first R partitions will have 1 more data point aggregated. 
For example, If there are 100 points and the count is 8, since remainder is 4, first 4 intervals will have 100 / 8 + 1 = 13 data points aggregated, the rest 4 intervals will have 12 data points aggregated, 8 data points will get returned.

Reply:
hmmm, interesting. That would mean the buckets are not the same size. Depending on what caller does with this - perhaps visualize as a graph? - that might be confusing.

Reply:
let's talk to @ajainarayanan to see how this is being used.